cbdr/cortex
# AR-66: Broken Pagination Headers

## Purpose:
We've always run into a wall when using `grape-kaminari` in combination with ElasticSearch queries. They didn't seem like they played well together, so we forked `grape-kaminari` and hacked together a method that would apply pagination headers manually. This approach proved brittle, however - these headers aren't applied consistently. I revisited our approach and dug into how the libraries could work together - and got it working, hurrah! We're no longer doing bespoke mumbo-jumbo, opting instead for `grape-kaminari`'s `paginate` helper method, which we now universally apply to our endpoints across Legacy. This approach still avoids grabbing the actual records from the database too early (which is the issue we always ran into in the past with `grape-kaminari`) - it does it only once, after all the pre-processing is complete.

*Edit:* The real solution was actually totally different, but the above changes are still good refactors. To fix consumer's pagination headers issue, we now cache the headers along with the Post records, and restore the cached headers with each response.

## JIRA:
https://cb-content-enablement.atlassian.net/browse/AR-66

## Steps to Take On Prod
N/A

## Changes:
* Changes to setup
  * N/A

* Architectural changes
  * N/A

* Migrations
  * N/A

* Library changes
  * Our forked `grape-kaminari` is no longer necessary

* Side effects
  * Each endpoint should be tested thoroughly prior to release. This could have unintended consequences. Luckily, this is one of the few parts of the application we *do* have some test coverage around.

## Screenshots
* Before
N/A

* After
N/A

## QA Links:
http://web.cortex-4.development.c66.me/

## How to Verify These Changes
* Specific pages to visit
  * N/A

* Steps to take
  * The easiest way to test is to upload a bunch of Media, such that there are 2 pages of it. Alternatively, you can modify the URL so that the total requested records per page is 1 or 2.
  * Ensure both the Admin Interface as well as the API Feed endpoints return the correct results, in the correct order, with the correct pagination headers

* Responsive considerations
  * N/A

## Relevant PRs/Dependencies:
N/A

## Additional Information
* @Casao has been looking out for this - heads up!
